### PR TITLE
Correction to get correct subscription policy from SubscriptionDataStore

### PR DESCRIPTION
**Pull request by slahirucd7**
$subject
Related issue: https://github.com/wso2-enterprise/choreo/issues/26811

### Issues

Fixes #

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 1]
```diff
@@ -85,14 +85,19 @@ public class JWTAuthenticator implements Authenticator {
     private final boolean isGatewayTokenCacheEnabled;
     private AbstractAPIMgtGatewayJWTGenerator jwtGenerator;
     private static final Set<String> prodTokenNonProdAllowedOrgs = new HashSet<>();
+    private static final String orgList = System.getenv("CUSTOM_SUBSCRIPTION_POLICY_HANDLING_ORG");
+    private static Set<String> orgSet = null;
```
Comment (malinthaprasan) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:89]: We can initialize this to an empty set. Then no need to do any null checks

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 2]
```diff
@@ -332,27 +337,27 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
+                            String metaDataOrgId = StringUtils.isNotEmpty(subPolicy.getOrganization()) &&
+                                    StringUtils.isNotEmpty(orgList) &&
+                                    (orgSet.contains(subPolicy.getOrganization()) || orgList.equals("*")) ?
+                                    subPolicy.getOrganization() : APIConstants.SUPER_TENANT_DOMAIN_NAME;
+                            log.debug("Subscription rate-limiting will be evaluated for the organization: " +
+                                    metaDataOrgId);
+                            requestContext.addMetadataToMap("ratelimit:organization", metaDataOrgId);
                         } else {
                             requestContext.addMetadataToMap("ratelimit:organization",
                                     APIConstants.SUPER_TENANT_DOMAIN_NAME);
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:355]: Does the value of `APIConstants.SUPER_TENANT_DOMAIN_NAME` is 'carbon.super'?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java [hunk 3]
```diff
@@ -546,7 +546,8 @@ public List<SubscriptionPolicy> getMatchingSubscriptionPolicies(String policyNam
         if (StringUtils.isEmpty(policyName)) {
             subscriptionPolicies.addAll(this.subscriptionPolicyMap.values());
         } else {
-            SubscriptionPolicy policy = this.getSubscriptionPolicyByName(policyName);
+            SubscriptionPolicy policy = this.getSubscriptionPolicyByOrgIdAndName(APIConstants.SUPER_TENANT_DOMAIN_NAME,
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java:549]: Why are we hardcoding SUPER_TENANT_DOMAIN_NAME here?

Comment (malinthaprasan) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java:549]: In other words:
- Where is this method being used? `getMatchingSubscriptionPolicies `
- Is it correct to assume the org is always SUPER_TENANT_DOMAIN_NAME when calling this method? i.e Why don't we need to pass orgId from a method variable?

Comment (slahirucd7) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java:549]: This method is relevant to the retrieval of subscription policies that are keeping in the Enforcer. As per the team discussions this was added during the JMS based event retrieval.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/models/SubscriptionPolicy.java [hunk 4]
```diff
@@ -78,17 +80,18 @@ public void setGraphQLMaxDepth(int graphQLMaxDepth) {
     }
     @Override
     public String getCacheKey() {
-
-        return PolicyType.SUBSCRIPTION + SubscriptionDataStoreUtil.getPolicyCacheKey(getName());
-
+        String org = StringUtils.isNotEmpty(getOrganization()) ? getOrganization() :
+                APIConstants.SUPER_TENANT_DOMAIN_NAME;
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/models/SubscriptionPolicy.java:84]: What if we update the getOrganization() method this logic?
So here we can just getOrganization() and inside getOrganization we will return SUPER_TENANT_DOMAIN_NAME if org is empty.

```suggestion
        String org = getOrganization()
```

Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/models/SubscriptionPolicy.java:84]: Is it possible to org be empty?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 5]
```diff
@@ -332,27 +337,27 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
-                                requestContext.addMetadataToMap("ratelimit:organization",
-                                        APIConstants.SUPER_TENANT_DOMAIN_NAME);
-                            }
+
+                        String matchedApiOrganizationId = requestContext.getMatchedAPI().getOrganizationId();
+                        if (datastore.getSubscriptionPolicyByOrgIdAndName(matchedApiOrganizationId, subPolicyName)
+                                != null) {
+                            SubscriptionPolicy subPolicy = datastore.getSubscriptionPolicyByOrgIdAndName
+                                    (matchedApiOrganizationId, subPolicyName);
+                            String metaDataOrgId = StringUtils.isNotEmpty(subPolicy.getOrganization()) &&
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:346]: We can remove this check StringUtils.isNotEmpty(subPolicy.getOrganization()), if we update the getOrganization() to return orgID or carbon.super as suggested in https://github.com/wso2/product-microgateway/pull/3512/files#r1601739840

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java (hunk 3 above)
Comment (renuka-fernando) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/subscription/SubscriptionDataStoreImpl.java:549]: @slahirucd7, this looks like a simple fix, Shall we do this?

We can add a new query param key, `organization` and extract it as here we've done for the policy name.

https://github.com/wso2/product-microgateway/blob/8371e29172a93a44dcd6fad28dcdab7320451d23/enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/admin/handlers/ThrottlingPolicyRequestHandler.java#L38-L47

/throttling_policies/application?name=Gold&orgnization=123456789

Then let's add the org param to this function

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/models/SubscriptionPolicy.java (hunk 4 above)
Comment (slahirucd7) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/models/SubscriptionPolicy.java:84]: We discussed this earlier, too. Integration tests don't have an org id.

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20240516.12

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240516.12

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240516.12

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20240516.12

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240517.4

**Comment by choreo-cicd** [comment]
[] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240520.2

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : stage-deployment-v2 : 20240520.2

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240520.2

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240520.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240521.2

**Comment by choreo-cicd** [comment]
[failed] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20240521.2

**Comment by choreo-cicd** [comment]
[canceled] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240521.2

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240521.3

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240521.3

**Pull request merged**